gax/src/main/java/com/google/api/gax/batching/FlowController.java

```diff
@@ -188,20 +247,31 @@ public void reserve(long elements, long bytes) throws FlowControlException {
 
     if (outstandingElementCount != null) {
       if (!outstandingElementCount.acquire(elements)) {
-        throw new MaxOutstandingElementCountReachedException(maxOutstandingElementCount);
+        throw new MaxOutstandingElementCountReachedException(currentOutstandingElementCount);
       }
     }
 
-    // Will always allow to send a request even if it is larger than the flow control limit,
+    // Always allows to send a request even if it is larger than the flow control limit,
     // if it doesn't then it will deadlock the thread.
     if (outstandingByteCount != null) {
-      long permitsToDraw = Math.min(bytes, maxOutstandingRequestBytes);
-      if (!outstandingByteCount.acquire(permitsToDraw)) {
+      long permitsToDraw, permitsOwed;
+      boolean acquired;
+      synchronized (this) {
+        // This needs to be synchronized so currentOutstandingRequestBytes won't get updated to a
+        // smaller value before calling acquire. Otherwise it'll cause deadlocks.
+        permitsToDraw = Math.min(bytes, currentOutstandingRequestBytes);
+        permitsOwed = bytes - permitsToDraw;
+        acquired = outstandingByteCount.acquire(permitsToDraw);
```

Comment:
I think the comment about avoiding deadlock above is misleading and led to a bug here.
I'm guessing the original intention was to allow oversized requests to be sent if they were larger than the max allowed byte size of flow control. In other words if maxOutstandingRequestBytes was 10 and the first request was 100 bytes, then it would never be sent. 
So I think what this code originally tried to do is allow the large request to be sent. So I think the correct behavior here should only acquire the partial tokens if permitsToDraw > 0

Reply:
Oh boy, it seems like the original implementation was a bit flawed:
it would reserve the diminished amount, but would release the original amount. 
I dont have the full understanding of the repercussions, so lets chat offline

Reply:
Rename currentOutstandingRequestBytes -> currentLimitRequestBytes